---
title: "isoviz"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{isoviz}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(isoviz)
```

# Introduction

Our package aims to help quantify and visualize the transcript isoforms that are present in your samples.

# Exemplary workflow

# Step 1: Formatting Input, transcript coords

We start off by loading our exon and intron coordinates for all transcripts. By default, this will use 'gencode.v41.basic annotation' data. You can also input your own psl file (or you can use a gtf and convert it to psl first with the function "isoviz_gtf_to_psl" and use its output as input for "isoviz_coords").

```{r}
# if you want to use our pre-loaded gencode v41 basic annotations, use this code
file_path <- system.file("data", "gencode.v41.basic.annotation.psl", package="isoviz")
gene_trans <- system.file("data", "gencode_v41_gene_transcript_convert.txt", package="isoviz")

all_coordinates <- isoviz_coords(file_path, gene_trans) #use default genome .psl file  

# if you want to use your own psl or gtf file (including output from a long read sequencing experiment), use this code
#file_path <- "" # input your path here for the gtf/psl file
#gene_trans <- system.file("data", "gencode_v41_gene_transcript_convert.txt", package="isoviz") # you can still use this file, transcripts that don't match gencode 41 basics will be marked as novel
#count_file <- "" # this parameter is options, but if you have a file with read counts per transcript from a long read sequencing experiment, you can input the path here. Expected format is two columns, tab-delimited, with trans.id_gene.id and readcounts columns, no header

#all_coordinates <- isoviz_coords(file_path, gene_trans, input_type="psl", count_file, min_count = 5)  

# Let's extract the exon coordinates 
exon_coords <- all_coordinates[[1]]

# And now the intron coordinates 
intron_coords <- all_coordinates[[2]]
```

Note: we can also use the intron coordinates from above to generate a mapping between junctions and trancsripts as shown below:
```{r}
uniqueness_summary = isoviz_junction_to_transcript_summary(intron_coords)
head(uniqueness_summary)
```

# Step 2: Formatting Input, short read counts with LeafCutter

We will first need to load our leafcutter junctions that we obtained by running Regtools extract junctions on our BAM files. You can use one of our preloaded cell types or first run this on your own BAM file and then use as input for this function. We will look at junctions in hESC data. We will first need to run 'minicutter' to cluster the junction coordinates and obtain intron cluster events.

```{r}
# load junctions for cell type of interest or input your own
junctions <- system.file("data", "hESC.junc.txt", package="isoviz")

# junction files also available for A375 and HEK293
# junctions <- system.file("data", "A375.junc.txt", package="isoviz")
# junctions <- system.file("data", "HEK293.junc.txt", package="isoviz")

# run minicutter to get clusters 
intron_clusts <- isoviz_minicutter(juncs_file = junctions)
```

# This reads in a table we created for gencode v41, providing additional junction information like uniqueness

Let's take a look at how junctions were grouped into intron clusters. Note, some of these clusters will only contain one junction (singleton). You will have the option to change this when running isoviz_minicutter.

```{r}
# For now we will use the expanded intron dataset with additional annotations by Megan 
# We will need an additional function to add these annotations 
intron_annotations <- system.file("data", "gencode_intron_all_data.rda", package="isoviz")
load(intron_annotations)
```

Step 3: Overlap exon/intron coords with leafcutter counts for a specific gene Let's work with RBFOX2. Now we have to map the observed junctions with their corresponding exons and transcript isoforms. To do this, we will use the `isoviz_map_junctions` function. Make sure to check strand of gene!

```{r}
gene = "RBFOX2"
gene_exons <- filter(exon_coords, gene_name == gene)
gene_introns <- filter(intron_coords, gene_name == gene)

mapped_junctions = isoviz_map_junctions(cell_type = "hESC", gene_introns, intron_clusts, gencode_intron_all_data)

# let's rescale the intron coordinates but keep the relative exon alignments 
# this step will return an updated dataframe of exon coordinates that we can use 
# with our plotting function

#rescaled_coords = isoviz_rescale_introns(gene_introns, gene_exons, width_rescale=10) 
```

Step 4: Make a plot!

```{r}
isoviz_plot_juncs_to_iso(mapped_junctions, gene_exons, gene_introns,
                                    cell_type = "hESC",
                                    junc_usage = 5, #min junc usage to be included 
                                    intron_scale = "no")
```

Now let's make the same plot but with scaled introns so that we can see the exons better! Note, this currently only works for junctions that we are able to annotate to transcript isoforms in the top plot!

```{r}
isoviz_plot_juncs_to_iso(mapped_junctions, gene_exons, gene_introns,
                                    cell_type = "hESC",
                                    junc_usage = 5, #min junc usage to be included 
                                    intron_scale = "yes", intron_scale_width = 10)
```

```{r}
isoviz_plot_juncs_to_iso(mapped_junctions, gene_exons, gene_introns,
                                    cell_type = "hESC",
                                    junc_usage = 5, #min junc usage to be included 
                                    intron_scale = "yes", intron_scale_width = 10, 
                                    include_specific_junctions = c("junc178147", "junc178149"))
#date <- format(Sys.Date(), "%y%m%d")
#ggsave(paste0("~/cas13-isoforms/isoviz_plots/", date, "_", gene,".png"), width = 8, height = 3)

```

```{r}
# RBFOX2
isoviz_plot_juncs_to_iso(mapped_junctions, gene_exons, gene_introns,
                                    cell_type = "hESC",
                                    junc_usage = 5, #min junc usage to be included 
                                    intron_scale = "yes", intron_scale_width = 10,
                                    include_specific_isoforms = c("RBFOX2-209", "RBFOX2-220", "RBFOX2-208", "RBFOX2-205"),
                                    include_specific_junctions = c("junc178147", "junc178149", "junc178135", "junc178136", "junc178145", "junc178146"))
#date <- format(Sys.Date(), "%y%m%d")
#ggsave(paste0("~/isoviz/plots/", date, "_", gene,".png"), width = 8, height = 2)

# COP1
# isoviz_plot_juncs_to_iso(mapped_junctions, gene_exons, gene_introns,
#                                     cell_type = "hESC",
#                                     junc_usage = 0, #min junc usage to be included 
#                                     intron_scale = "yes", intron_scale_width = 10,
#                                     include_specific_isoforms = c("COP1-201", "COP1-204"),
#                                     include_specific_junctions = c("junc20760", "junc20769", "junc20759"))
#date <- format(Sys.Date(), "%y%m%d")
#ggsave(paste0("~/isoviz/plots/", date, "_", gene,".pdf"), width = 8, height = 3)

# SPAG9
# isoviz_plot_juncs_to_iso(mapped_junctions, gene_exons, gene_introns,
#                                     cell_type = "hESC",
#                                     junc_usage = 0, #min junc usage to be included 
#                                     intron_scale = "yes", intron_scale_width = 10,
#                                     include_specific_junctions = c("junc117123", "junc117141", "junc117144"))
# #date <- format(Sys.Date(), "%y%m%d")
# #ggsave(paste0("~/isoviz/plots/", date, "_", gene,".pdf"), width = 8, height = 3)
# 
# # MKNK2
# isoviz_plot_juncs_to_iso(mapped_junctions, gene_exons, gene_introns,
#                                     cell_type = "hESC",
#                                     junc_usage = 0, #min junc usage to be included 
#                                     intron_scale = "yes", intron_scale_width = 10,
#                                     include_specific_isoforms = c("MKNK2-201", "MKNK2-202"),
#                                     include_specific_junctions = c("junc128207", "junc128208"))
# #date <- format(Sys.Date(), "%y%m%d")
#ggsave(paste0("~/isoviz/plots/", date, "_", gene,".pdf"), width = 8, height = 2.5)

```

Step 5: Guide Predictions Finally, if you are interested in obtaining a list of gRNAs that you could use to target specific junctions with Cas13, you can do the following:

```{r}
#library(flextable)

guide_table = isoviz_get_guide_predictions(gene = "ENSG00000100320", leafcutter_input=intron_clusts,
                                           guides_per_junction = 5, 
                                           include_specific_junctions = c("junc178147", "junc178149", "junc178135", "junc178136", "junc178145", "junc178146"), output_format="nice_table")

guide_table

# guide_table = isoviz_get_guide_predictions(gene = "ENSG00000100320", leafcutter_input=intron_clusts,
#                                            guides_per_junction = 8,
#                                            include_specific_junctions = c("junc178149"))
# 
# # for paper
# guide_table = isoviz_get_guide_predictions(gene = "ENSG00000100320", leafcutter_input=intron_clusts,
#                                            guides_per_junction = 5, 
#                                            include_specific_junctions = c("junc178147", "junc178149", "junc178135", "junc178136", "junc178145", "junc178146"))
# 
# class(guide_table)
# 

#date <- format(Sys.Date(), "%y%m%d")
#save_as_image(x = guide_table, path = paste0("~/cas13-isoforms/isoviz_plots/", date, "_table_", gene,".pdf"), res = 300)

```
