---
title: "isoviz"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{isoviz}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(isoviz)
```

# Introduction

Our package aims to help quantify and visualize the transcript isoforms that are present in your samples. 

# Exemplary workflow 

We start off by loading all our exon and intron coordinates. By default, this will use 'gencode.v41.basic annotation' data. You can also input your own psl file (gtf file as well in the future).  
```{r}

file_path <- system.file("data", "datasources/gencode.v41.basic.annotation.psl", package = "isoviz")

all_coordinates = isoviz_coords(file_path,
                         input_type="psl") #use default genome .psl file  
```

```{r}
# Let's look at the exon coordinates 
exon_coords = all_coordinates[[1]]

# Let's look at the intron coordinates 
intron_coords = all_coordinates[[1]]
```

Note that the exon coordinates will be used to for plotting transcript isoforms and the intron coordinates will be used in conjunction with exon-exon junctions and intron cluster coordinates from leafcutter analysis. 
```{r}
char_summary( month.name )
```
