---
title: "isoviz"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{isoviz}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(isoviz)
```

# Introduction

Our package aims to help quantify and visualize the transcript isoforms that are present in your samples.

# Exemplary workflow 

We start off by loading all our exon and intron coordinates. By default, this will use 'gencode.v41.basic annotation' data. You can also input your own psl file (gtf file as well in the future).  

```{r}
file_path <- system.file("data", "gencode.v41.basic.annotation.psl", package="isoviz")
#gene_trans <- system.file("data", "gencode_v41_gene_transcript_convert", package="isoviz")

#all_coordinates = isoviz_coords(file_path, gene_trans, input_type="psl") #use default genome .psl file  
```

```{r}
# Let's look at the exon coordinates 
#exon_coords = all_coordinates[[1]]

# Let's look at the intron coordinates 
#intron_coords = all_coordinates[[1]]
```


