<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Visualize transcript isoforms alongside exon-exon junction counts • isoviz</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Visualize transcript isoforms alongside exon-exon junction counts">
<meta property="og:description" content="Visualize which junctions are being utilized alongside annotated
    transcript isoforms. Find appropriate guide RNAs to target junction for 
    isoform specific knockdown. ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">isoviz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="articles/isoviz.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="isoviz-">isoviz <img src="inst/figures/logo.png" align="right" height="139"><a class="anchor" aria-label="anchor" href="#isoviz-"></a>
</h1></div>
<!-- badges: start -->

<p><code>isoviz</code> is a package that allows for simplified transcript isoforms visualizations.</p>
<p>The goal of isoviz is to simplify working with trancsript isoforms. Given a gene name or Ensembl ID, you can visualize all transcript structures. Additionally, Isoviz integrates with junction coordinates and counts obtained with Regtools. You can choose to visualize only the transcript isoforms that are detected in your sample and see how they group into leafuctter intron clusters.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the development version of isoviz from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"daklab/isoviz"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Our package aims to help quantify and visualize the transcript isoforms that are present in your samples.</p>
</div>
<div class="section level2">
<h2 id="example">Example<a class="anchor" aria-label="anchor" href="#example"></a>
</h2>
<p>We start off by loading all our exon and intron coordinates. By default, this will use ‘gencode.v41.basic annotation’ data. You can also input your own psl file (gtf file as well in the future).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">isoviz</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## basic example code</span></span>
<span><span class="va">file_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"gencode.v41.basic.annotation.psl"</span>, package<span class="op">=</span><span class="st">"isoviz"</span><span class="op">)</span></span>
<span><span class="va">gene_trans</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"gencode_v41_gene_transcript_convert.txt"</span>, package<span class="op">=</span><span class="st">"isoviz"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">all_coordinates</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/isoviz_coords.html">isoviz_coords</a></span><span class="op">(</span><span class="va">file_path</span>, <span class="va">gene_trans</span>, input_type<span class="op">=</span><span class="st">"psl"</span><span class="op">)</span> <span class="co">#use default genome .psl file  </span></span></code></pre></div>
<div class="section level4">
<h4 id="lets-look-at-the-exon-coordinates">Let’s look at the exon coordinates<a class="anchor" aria-label="anchor" href="#lets-look-at-the-exon-coordinates"></a>
</h4>
<pre class="{r}"><code><span><span class="va">exon_coords</span> <span class="op">&lt;-</span> <span class="va">all_coordinates</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">exon_coords</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="lets-look-at-the-intron-coordinates">Let’s look at the intron coordinates<a class="anchor" aria-label="anchor" href="#lets-look-at-the-intron-coordinates"></a>
</h4>
<pre class="{r}"><code><span><span class="va">intron_coords</span> <span class="op">&lt;-</span> <span class="va">all_coordinates</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">intron_coords</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="lets-make-a-basic-plot-of-the-isoforms-for-the-gene-rbfox2">Let’s make a basic plot of the isoforms for the gene RBFOX2<a class="anchor" aria-label="anchor" href="#lets-make-a-basic-plot-of-the-isoforms-for-the-gene-rbfox2"></a>
</h4>
<pre class="{r}"><code><span><span class="co"># get exons and introns </span></span>
<span><span class="va">rbfox2_exons</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">exon_coords</span>, <span class="va">gene_name</span> <span class="op">==</span> <span class="st">"RBFOX2"</span><span class="op">)</span></span>
<span><span class="va">rbfox2_introns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">intron_coords</span>, <span class="va">gene_name</span> <span class="op">==</span> <span class="st">"RBFOX2"</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="intron-re-scaling">Intron re-scaling<a class="anchor" aria-label="anchor" href="#intron-re-scaling"></a>
</h4>
<p>Let’s rescale the intron coordinates but keep the relative exon alignments this step will return an updated dataframe of exon coordinates that we can use with our plotting function.</p>
<pre class="{r}"><code><span><span class="va">rescaled_coords</span> <span class="op">=</span> <span class="fu"><a href="reference/isoviz_rescale_introns.html">isoviz_rescale_introns</a></span><span class="op">(</span><span class="va">rbfox2_introns</span>, <span class="va">rbfox2_exons</span>, width_rescale<span class="op">=</span><span class="fl">10</span><span class="op">)</span> </span></code></pre>
</div>
<div class="section level4">
<h4 id="intron-clustering-using-junction-count-data-from-bam-files">Intron clustering using junction count data from BAM files<a class="anchor" aria-label="anchor" href="#intron-clustering-using-junction-count-data-from-bam-files"></a>
</h4>
<p>We will first need to load our leafcutter junctions that we obtained by running Regtools extract junctions on our BAM files. You can use one of our preloaded cell types or first run this on your own BAM file and then use as input for this function. We will look at junctions in hESC data. We will first need to run ‘minicutter’ to cluster the junction coordinates and obtain intron cluster events.</p>
<pre class="{r}"><code><span></span>
<span><span class="co"># load junctions for cell type of interest or input your own</span></span>
<span><span class="va">junctions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"hESC-MKNK2-G1_v41_basic.junc"</span>, package<span class="op">=</span><span class="st">"isoviz"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># run minicutter to get clusters </span></span>
<span><span class="va">intron_clusts</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/isoviz_minicutter.html">isoviz_minicutter</a></span><span class="op">(</span>juncs_file <span class="op">=</span> <span class="va">junctions</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">intron_clusts</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<p>Let’s take a look at how junctions were grouped into intron clusters. Note, some of these clusters will only contain one junction (singleton). You will have the option to change this when running isoviz_minicutter.</p>
<pre class="{r}"><code><span><span class="co"># For now we will use the expanded intron dataset with additional annotations by Megan </span></span>
<span><span class="co"># We will need an additional function to add these annotations </span></span>
<span><span class="va">intron_annotations</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"gencode_intron_all_data.rda"</span>, package<span class="op">=</span><span class="st">"isoviz"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="va">intron_annotations</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="plotting-isoforms-and-junction-and-cluster-information">Plotting isoforms and junction and cluster information<a class="anchor" aria-label="anchor" href="#plotting-isoforms-and-junction-and-cluster-information"></a>
</h4>
<p>Let’s continue working on RBFOX2. Now we have to map the observed junctions with their corresponding exons and transcript isoforms. To do this, we will use the <code>isoviz_map_junctions</code> function. Make sure to check strand of gene!</p>
<pre class="{r}"><code><span><span class="va">mapped_junctions</span> <span class="op">=</span> <span class="fu"><a href="reference/isoviz_map_junctions.html">isoviz_map_junctions</a></span><span class="op">(</span>cell_type <span class="op">=</span> <span class="st">"hESC"</span>, <span class="va">rbfox2_introns</span>, <span class="va">intron_clusts</span>, <span class="va">gencode_intron_all_data</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mapped_junctions</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<p>Now we are ready to make a plot!</p>
<pre class="{r}"><code><span><span class="fu"><a href="reference/isoviz_plot_juncs_to_iso.html">isoviz_plot_juncs_to_iso</a></span><span class="op">(</span><span class="va">mapped_junctions</span>, <span class="va">rbfox2_exons</span>,</span>
<span>                                    cell_type <span class="op">=</span> <span class="st">"hESC"</span>,</span>
<span>                                    junction_usage <span class="op">=</span> <span class="fl">5</span>, <span class="co">#min junc usage to be included </span></span>
<span>                                    include_all_juncs <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                    include_specific_junctions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"JUNC00193787"</span>, <span class="st">"JUNC00193785"</span>, <span class="st">"JUNC00193786"</span>, <span class="st">"unk.7"</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span></code></pre>
<p><img src="inst/figures/Readme_RBFOX2.png" width="100%"></p>
</div>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/citation.html" class="external-link">citation</a></span><span class="op">(</span><span class="st">"isoviz"</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; To cite package ‘isoviz’ in publications use:</span></span>
<span><span class="co">#&gt;</span></span>
<span><span class="co">#&gt;  Isaev K, Schertzer M (2023). _isoviz: Visualize transcript isoforms alongside exon-exon junction counts_. R package version 0.1.0.</span></span>
<span><span class="co">#&gt;</span></span>
<span><span class="co">#&gt; A BibTeX entry for LaTeX users is</span></span>
<span><span class="co">#&gt;</span></span>
<span><span class="co">#&gt;  @Manual{,</span></span>
<span><span class="co">#&gt;    title = {isoviz: Visualize transcript isoforms alongside exon-exon junction counts},</span></span>
<span><span class="co">#&gt;    author = {Karin Isaev and Megan Schertzer},</span></span>
<span><span class="co">#&gt;    year = {2023},</span></span>
<span><span class="co">#&gt;    note = {R package version 0.1.0},</span></span>
<span><span class="co">#&gt;  }</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="questions-or-suggestions">Questions or suggestions?<a class="anchor" aria-label="anchor" href="#questions-or-suggestions"></a>
</h2>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing isoviz</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li> <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://img.shields.io/github/issues/daklab/isoviz/issues" class="external-link"><img src="https://img.shields.io/github/issues/daklab/isoviz" alt="GitHub issues"></a></li>
<li><a href="https://travis-ci.com/karini925/isoviz" class="external-link"><img src="https://travis-ci.com/karini925/isoviz.svg?branch=master" alt="Travis build status"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by .</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
