<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>isoviz • isoviz</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="isoviz">
<meta property="og:description" content="isoviz">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">isoviz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/isoviz.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>isoviz</h1>
            
      
      
      <div class="hidden name"><code>isoviz.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">isoviz</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: data.table</span></span>
<span><span class="co">#&gt; Loading required package: ggpubr</span></span>
<span><span class="co">#&gt; Loading required package: ggplot2</span></span>
<span><span class="co">#&gt; Loading required package: readr</span></span>
<span><span class="co">#&gt; Loading required package: tidyverse</span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">Attaching core tidyverse packages</span> ──────────────────────── tidyverse 2.0.0 ──</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">dplyr    </span> 1.1.2     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">stringr  </span> 1.5.0</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">forcats  </span> 1.0.0     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">tibble   </span> 3.2.1</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">lubridate</span> 1.9.2     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">tidyr    </span> 1.3.0</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">purrr    </span> 1.0.1     </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">Conflicts</span> ────────────────────────────────────────── tidyverse_conflicts() ──</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">dplyr</span>::<span style="color: #00BB00;">between()</span>     masks <span style="color: #0000BB;">data.table</span>::between()</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">dplyr</span>::<span style="color: #00BB00;">filter()</span>      masks <span style="color: #0000BB;">stats</span>::filter()</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">dplyr</span>::<span style="color: #00BB00;">first()</span>       masks <span style="color: #0000BB;">data.table</span>::first()</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">lubridate</span>::<span style="color: #00BB00;">hour()</span>    masks <span style="color: #0000BB;">data.table</span>::hour()</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">lubridate</span>::<span style="color: #00BB00;">isoweek()</span> masks <span style="color: #0000BB;">data.table</span>::isoweek()</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">dplyr</span>::<span style="color: #00BB00;">lag()</span>         masks <span style="color: #0000BB;">stats</span>::lag()</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">dplyr</span>::<span style="color: #00BB00;">last()</span>        masks <span style="color: #0000BB;">data.table</span>::last()</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">lubridate</span>::<span style="color: #00BB00;">mday()</span>    masks <span style="color: #0000BB;">data.table</span>::mday()</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">lubridate</span>::<span style="color: #00BB00;">minute()</span>  masks <span style="color: #0000BB;">data.table</span>::minute()</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">lubridate</span>::<span style="color: #00BB00;">month()</span>   masks <span style="color: #0000BB;">data.table</span>::month()</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">lubridate</span>::<span style="color: #00BB00;">quarter()</span> masks <span style="color: #0000BB;">data.table</span>::quarter()</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">lubridate</span>::<span style="color: #00BB00;">second()</span>  masks <span style="color: #0000BB;">data.table</span>::second()</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">purrr</span>::<span style="color: #00BB00;">transpose()</span>   masks <span style="color: #0000BB;">data.table</span>::transpose()</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">lubridate</span>::<span style="color: #00BB00;">wday()</span>    masks <span style="color: #0000BB;">data.table</span>::wday()</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">lubridate</span>::<span style="color: #00BB00;">week()</span>    masks <span style="color: #0000BB;">data.table</span>::week()</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">lubridate</span>::<span style="color: #00BB00;">yday()</span>    masks <span style="color: #0000BB;">data.table</span>::yday()</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">lubridate</span>::<span style="color: #00BB00;">year()</span>    masks <span style="color: #0000BB;">data.table</span>::year()</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Use the conflicted package (<span style="color: #0000BB; font-style: italic;">&lt;http://conflicted.r-lib.org/&gt;</span>) to force all conflicts to become errors</span></span>
<span><span class="co">#&gt; Loading required package: Matrix</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'Matrix'</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:tidyr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand, pack, unpack</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Loading required package: igraph</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'igraph'</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:lubridate':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     %--%, union</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:dplyr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     as_data_frame, groups, union</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:purrr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     compose, simplify</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; The following object is masked from 'package:tidyr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     crossing</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; The following object is masked from 'package:tibble':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     as_data_frame</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     decompose, spectrum</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; The following object is masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     union</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Loading required package: magrittr</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'magrittr'</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; The following object is masked from 'package:purrr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     set_names</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; The following object is masked from 'package:tidyr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     extract</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Loading required package: ggsci</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Loading required package: cowplot</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'cowplot'</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; The following object is masked from 'package:lubridate':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     stamp</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; The following object is masked from 'package:ggpubr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     get_legend</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Our package aims to help quantify and visualize the transcript
isoforms that are present in your samples.</p>
</div>
<div class="section level2">
<h2 id="exemplary-workflow">Exemplary workflow<a class="anchor" aria-label="anchor" href="#exemplary-workflow"></a>
</h2>
<p>We start off by loading all our exon and intron coordinates. By
default, this will use ‘gencode.v41.basic annotation’ data. You can also
input your own psl file (gtf file as well in the future).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">file_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"gencode.v41.basic.annotation.psl"</span>, package<span class="op">=</span><span class="st">"isoviz"</span><span class="op">)</span></span>
<span><span class="va">gene_trans</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"gencode_v41_gene_transcript_convert.txt"</span>, package<span class="op">=</span><span class="st">"isoviz"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">all_coordinates</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/isoviz_coords.html">isoviz_coords</a></span><span class="op">(</span><span class="va">file_path</span>, <span class="va">gene_trans</span>, input_type<span class="op">=</span><span class="st">"psl"</span><span class="op">)</span> <span class="co">#use default genome .psl file  </span></span>
<span><span class="co">#&gt; [1] "116566 transcripts with at least one intron!"</span></span>
<span><span class="co">#&gt; Adding missing grouping variables: `trans_id`</span></span>
<span><span class="co">#&gt; Adding missing grouping variables: `trans_id`</span></span>
<span><span class="co">#&gt; [1] "116566 transcripts with at least one intron!"</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">116649</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">6</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Delimiter:</span> "\t"</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">chr</span> (6): gene_id, trans_id, gene_name, transcript_name, gene_type, transcrip...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Let's look at the exon coordinates </span></span>
<span><span class="va">exon_coords</span> <span class="op">&lt;-</span> <span class="va">all_coordinates</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">exon_coords</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;     chr start   end          trans_id           gene_id strand blocksizes</span></span>
<span><span class="co">#&gt; 1: chr1 11868 14409 ENST00000456328.2 ENSG00000223972.5      +        359</span></span>
<span><span class="co">#&gt; 2: chr1 11868 14409 ENST00000456328.2 ENSG00000223972.5      +        109</span></span>
<span><span class="co">#&gt; 3: chr1 11868 14409 ENST00000456328.2 ENSG00000223972.5      +       1189</span></span>
<span><span class="co">#&gt; 4: chr1 12009 13670 ENST00000450305.2 ENSG00000223972.5      +         48</span></span>
<span><span class="co">#&gt; 5: chr1 12009 13670 ENST00000450305.2 ENSG00000223972.5      +         49</span></span>
<span><span class="co">#&gt; 6: chr1 12009 13670 ENST00000450305.2 ENSG00000223972.5      +         85</span></span>
<span><span class="co">#&gt;    blockstarts transcript_length blockends gene_name transcript_name</span></span>
<span><span class="co">#&gt; 1:       11868              2541     12227   DDX11L1     DDX11L1-202</span></span>
<span><span class="co">#&gt; 2:       12612              2541     12721   DDX11L1     DDX11L1-202</span></span>
<span><span class="co">#&gt; 3:       13220              2541     14409   DDX11L1     DDX11L1-202</span></span>
<span><span class="co">#&gt; 4:       12009              1661     12057   DDX11L1     DDX11L1-201</span></span>
<span><span class="co">#&gt; 5:       12178              1661     12227   DDX11L1     DDX11L1-201</span></span>
<span><span class="co">#&gt; 6:       12612              1661     12697   DDX11L1     DDX11L1-201</span></span>
<span><span class="co">#&gt;                             gene_type                    transcript_type</span></span>
<span><span class="co">#&gt; 1: transcribed_unprocessed_pseudogene               processed_transcript</span></span>
<span><span class="co">#&gt; 2: transcribed_unprocessed_pseudogene               processed_transcript</span></span>
<span><span class="co">#&gt; 3: transcribed_unprocessed_pseudogene               processed_transcript</span></span>
<span><span class="co">#&gt; 4: transcribed_unprocessed_pseudogene transcribed_unprocessed_pseudogene</span></span>
<span><span class="co">#&gt; 5: transcribed_unprocessed_pseudogene transcribed_unprocessed_pseudogene</span></span>
<span><span class="co">#&gt; 6: transcribed_unprocessed_pseudogene transcribed_unprocessed_pseudogene</span></span>
<span></span>
<span><span class="co"># Let's look at the intron coordinates </span></span>
<span><span class="va">intron_coords</span> <span class="op">&lt;-</span> <span class="va">all_coordinates</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">intron_coords</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;      chr intron_starts intron_ends            gene_id           trans_id strand</span></span>
<span><span class="co">#&gt; 1:  chr7     127588565   127589083 ENSG00000004059.11 ENST00000000233.10      +</span></span>
<span><span class="co">#&gt; 2:  chr7     127589163   127589485 ENSG00000004059.11 ENST00000000233.10      +</span></span>
<span><span class="co">#&gt; 3:  chr7     127589594   127590066 ENSG00000004059.11 ENST00000000233.10      +</span></span>
<span><span class="co">#&gt; 4:  chr7     127590137   127590963 ENSG00000004059.11 ENST00000000233.10      +</span></span>
<span><span class="co">#&gt; 5:  chr7     127591088   127591213 ENSG00000004059.11 ENST00000000233.10      +</span></span>
<span><span class="co">#&gt; 6: chr12       8941940     8942416  ENSG00000003056.8  ENST00000000412.8      -</span></span>
<span><span class="co">#&gt;    gene_name transcript_name      gene_type transcript_type</span></span>
<span><span class="co">#&gt; 1:      ARF5        ARF5-201 protein_coding  protein_coding</span></span>
<span><span class="co">#&gt; 2:      ARF5        ARF5-201 protein_coding  protein_coding</span></span>
<span><span class="co">#&gt; 3:      ARF5        ARF5-201 protein_coding  protein_coding</span></span>
<span><span class="co">#&gt; 4:      ARF5        ARF5-201 protein_coding  protein_coding</span></span>
<span><span class="co">#&gt; 5:      ARF5        ARF5-201 protein_coding  protein_coding</span></span>
<span><span class="co">#&gt; 6:      M6PR        M6PR-201 protein_coding  protein_coding</span></span></code></pre></div>
<p>Let’s make a basic plot of the isoforms for the gene RBFOX2. We will
also look at how to rescale the coordinates in case the gene has very
long introns.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rbfox2_exons</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">exon_coords</span>, <span class="va">gene_name</span> <span class="op">==</span> <span class="st">"PTEN"</span><span class="op">)</span></span>
<span><span class="va">rbfox2_introns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">intron_coords</span>, <span class="va">gene_name</span> <span class="op">==</span> <span class="st">"PTEN"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># let's rescale the intron coordinates but keep the relative exon alignments </span></span>
<span><span class="co"># this step will return an updated dataframe of exon coordinates that we can use </span></span>
<span><span class="co"># with our plotting function</span></span>
<span></span>
<span><span class="va">rescaled_coords</span> <span class="op">=</span> <span class="fu"><a href="../reference/isoviz_rescale_introns.html">isoviz_rescale_introns</a></span><span class="op">(</span><span class="va">rbfox2_introns</span>, <span class="va">rbfox2_exons</span>, width_rescale<span class="op">=</span><span class="fl">10</span><span class="op">)</span> </span>
<span><span class="co">#&gt; [1] "+"</span></span>
<span><span class="co">#&gt; [1] "PTEN-209"</span></span>
<span><span class="co">#&gt; [1] "getting ref coordinates"</span></span>
<span><span class="co">#&gt; [1] "PTEN-201"</span></span>
<span><span class="co">#&gt; [1] "PTEN-204"</span></span>
<span><span class="co">#&gt; [1] "PTEN-211"</span></span>
<span><span class="co">#&gt; [1] "PTEN-212"</span></span>
<span><span class="co">#&gt; [1] 1</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">rescaled_coords</span><span class="op">)</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt;     transcript_name   chr new_e_start new_e_end strand    start      end</span></span>
<span><span class="co">#&gt;  1:        PTEN-201 chr10      -218.9     705.1      + 87863624 87864548</span></span>
<span><span class="co">#&gt;  2:        PTEN-201 chr10      3652.8    3737.8      + 87894024 87894109</span></span>
<span><span class="co">#&gt;  3:        PTEN-201 chr10      6878.2    6923.2      + 87925512 87925557</span></span>
<span><span class="co">#&gt;  4:        PTEN-201 chr10      7472.1    7516.1      + 87931045 87931089</span></span>
<span><span class="co">#&gt;  5:        PTEN-201 chr10      7708.5    7947.5      + 87933012 87933251</span></span>
<span><span class="co">#&gt;  6:        PTEN-201 chr10      9834.2    9976.2      + 87952117 87952259</span></span>
<span><span class="co">#&gt;  7:        PTEN-201 chr10     10535.6   10702.6      + 87957852 87958019</span></span>
<span><span class="co">#&gt;  8:        PTEN-201 chr10     10990.1   11215.1      + 87960893 87961118</span></span>
<span><span class="co">#&gt;  9:        PTEN-201 chr10     11632.0   18276.0      + 87965286 87971930</span></span>
<span><span class="co">#&gt; 10:        PTEN-204 chr10       485.1     705.1      + 87864328 87864548</span></span>
<span><span class="co">#&gt; 11:        PTEN-204 chr10      3652.8    3737.8      + 87894024 87894109</span></span>
<span><span class="co">#&gt; 12:        PTEN-204 chr10      6878.2    6923.2      + 87925512 87925557</span></span>
<span><span class="co">#&gt; 13:        PTEN-204 chr10      7472.1    7516.1      + 87931045 87931089</span></span>
<span><span class="co">#&gt; 14:        PTEN-204 chr10      7708.5    7947.5      + 87933012 87933251</span></span>
<span><span class="co">#&gt; 15:        PTEN-204 chr10      9834.2    9976.2      + 87952117 87952259</span></span>
<span><span class="co">#&gt; 16:        PTEN-204 chr10     10535.6   10702.6      + 87957852 87958019</span></span>
<span><span class="co">#&gt; 17:        PTEN-204 chr10     10990.1   11406.1      + 87960893 87961309</span></span>
<span><span class="co">#&gt; 18:        PTEN-209 chr10         0.0     476.0      + 87862637 87863113</span></span>
<span><span class="co">#&gt; 19:        PTEN-209 chr10       610.1     705.1      + 87864453 87864548</span></span>
<span><span class="co">#&gt; 20:        PTEN-209 chr10      3652.8    3737.8      + 87894024 87894109</span></span>
<span><span class="co">#&gt; 21:        PTEN-209 chr10      6878.2    6923.2      + 87925512 87925557</span></span>
<span><span class="co">#&gt; 22:        PTEN-209 chr10      7472.1    7516.1      + 87931045 87931089</span></span>
<span><span class="co">#&gt; 23:        PTEN-209 chr10      7708.5    7947.5      + 87933012 87933251</span></span>
<span><span class="co">#&gt; 24:        PTEN-209 chr10      9834.2    9976.2      + 87952117 87952259</span></span>
<span><span class="co">#&gt; 25:        PTEN-209 chr10     10535.6   10702.6      + 87957852 87958019</span></span>
<span><span class="co">#&gt; 26:        PTEN-209 chr10     10990.1   11215.1      + 87960893 87961118</span></span>
<span><span class="co">#&gt; 27:        PTEN-209 chr10     11632.0   13231.0      + 87965286 87966885</span></span>
<span><span class="co">#&gt; 28:        PTEN-211 chr10       261.1     705.1      + 87864104 87864548</span></span>
<span><span class="co">#&gt; 29:        PTEN-211 chr10      3652.8    3737.8      + 87894024 87894109</span></span>
<span><span class="co">#&gt; 30:        PTEN-211 chr10      6878.2    6923.2      + 87925512 87925557</span></span>
<span><span class="co">#&gt; 31:        PTEN-211 chr10      7472.1    7516.1      + 87931045 87931089</span></span>
<span><span class="co">#&gt; 32:        PTEN-211 chr10      7708.5    7947.5      + 87933012 87933251</span></span>
<span><span class="co">#&gt; 33:        PTEN-211 chr10      9834.2    9976.2      + 87952117 87952259</span></span>
<span><span class="co">#&gt; 34:        PTEN-211 chr10     10535.6   10702.6      + 87957852 87958019</span></span>
<span><span class="co">#&gt; 35:        PTEN-211 chr10     10990.1   11215.1      + 87960893 87961118</span></span>
<span><span class="co">#&gt; 36:        PTEN-211 chr10     11632.0   18276.0      + 87965286 87971930</span></span>
<span><span class="co">#&gt; 37:        PTEN-211 chr10      -405.1     260.9      + 87863437 87864103</span></span>
<span><span class="co">#&gt; 38:        PTEN-212 chr10       -48.9     705.1      + 87863794 87864548</span></span>
<span><span class="co">#&gt; 39:        PTEN-212 chr10      3652.8    3737.8      + 87894024 87894109</span></span>
<span><span class="co">#&gt; 40:        PTEN-212 chr10      7472.1    7516.1      + 87931045 87931089</span></span>
<span><span class="co">#&gt; 41:        PTEN-212 chr10      7708.5    7947.5      + 87933012 87933251</span></span>
<span><span class="co">#&gt; 42:        PTEN-212 chr10      9834.2    9976.2      + 87952117 87952259</span></span>
<span><span class="co">#&gt; 43:        PTEN-212 chr10     10535.6   10702.6      + 87957852 87958019</span></span>
<span><span class="co">#&gt; 44:        PTEN-212 chr10     10990.1   11215.1      + 87960893 87961118</span></span>
<span><span class="co">#&gt; 45:        PTEN-212 chr10     11632.0   13231.0      + 87965286 87966885</span></span>
<span><span class="co">#&gt;     transcript_name   chr new_e_start new_e_end strand    start      end</span></span>
<span><span class="co">#&gt;     blocksizes exon_num min_start max_end transcript_length</span></span>
<span><span class="co">#&gt;  1:        924        1    -218.9 18276.0           18494.9</span></span>
<span><span class="co">#&gt;  2:         85        2    -218.9 18276.0           18494.9</span></span>
<span><span class="co">#&gt;  3:         45        3    -218.9 18276.0           18494.9</span></span>
<span><span class="co">#&gt;  4:         44        4    -218.9 18276.0           18494.9</span></span>
<span><span class="co">#&gt;  5:        239        5    -218.9 18276.0           18494.9</span></span>
<span><span class="co">#&gt;  6:        142        6    -218.9 18276.0           18494.9</span></span>
<span><span class="co">#&gt;  7:        167        7    -218.9 18276.0           18494.9</span></span>
<span><span class="co">#&gt;  8:        225        8    -218.9 18276.0           18494.9</span></span>
<span><span class="co">#&gt;  9:       6644        9    -218.9 18276.0           18494.9</span></span>
<span><span class="co">#&gt; 10:        220        1     485.1 11406.1           10921.0</span></span>
<span><span class="co">#&gt; 11:         85        2     485.1 11406.1           10921.0</span></span>
<span><span class="co">#&gt; 12:         45        3     485.1 11406.1           10921.0</span></span>
<span><span class="co">#&gt; 13:         44        4     485.1 11406.1           10921.0</span></span>
<span><span class="co">#&gt; 14:        239        5     485.1 11406.1           10921.0</span></span>
<span><span class="co">#&gt; 15:        142        6     485.1 11406.1           10921.0</span></span>
<span><span class="co">#&gt; 16:        167        7     485.1 11406.1           10921.0</span></span>
<span><span class="co">#&gt; 17:        416        8     485.1 11406.1           10921.0</span></span>
<span><span class="co">#&gt; 18:        476        1       0.0 13231.0           13231.0</span></span>
<span><span class="co">#&gt; 19:         95        2       0.0 13231.0           13231.0</span></span>
<span><span class="co">#&gt; 20:         85        3       0.0 13231.0           13231.0</span></span>
<span><span class="co">#&gt; 21:         45        4       0.0 13231.0           13231.0</span></span>
<span><span class="co">#&gt; 22:         44        5       0.0 13231.0           13231.0</span></span>
<span><span class="co">#&gt; 23:        239        6       0.0 13231.0           13231.0</span></span>
<span><span class="co">#&gt; 24:        142        7       0.0 13231.0           13231.0</span></span>
<span><span class="co">#&gt; 25:        167        8       0.0 13231.0           13231.0</span></span>
<span><span class="co">#&gt; 26:        225        9       0.0 13231.0           13231.0</span></span>
<span><span class="co">#&gt; 27:       1599       10       0.0 13231.0           13231.0</span></span>
<span><span class="co">#&gt; 28:        444        2    -405.1 18276.0           18681.1</span></span>
<span><span class="co">#&gt; 29:         85        3    -405.1 18276.0           18681.1</span></span>
<span><span class="co">#&gt; 30:         45        4    -405.1 18276.0           18681.1</span></span>
<span><span class="co">#&gt; 31:         44        5    -405.1 18276.0           18681.1</span></span>
<span><span class="co">#&gt; 32:        239        6    -405.1 18276.0           18681.1</span></span>
<span><span class="co">#&gt; 33:        142        7    -405.1 18276.0           18681.1</span></span>
<span><span class="co">#&gt; 34:        167        8    -405.1 18276.0           18681.1</span></span>
<span><span class="co">#&gt; 35:        225        9    -405.1 18276.0           18681.1</span></span>
<span><span class="co">#&gt; 36:       6644       10    -405.1 18276.0           18681.1</span></span>
<span><span class="co">#&gt; 37:        666        1    -405.1 18276.0           18681.1</span></span>
<span><span class="co">#&gt; 38:        754        1     -48.9 13231.0           13279.9</span></span>
<span><span class="co">#&gt; 39:         85        2     -48.9 13231.0           13279.9</span></span>
<span><span class="co">#&gt; 40:         44        3     -48.9 13231.0           13279.9</span></span>
<span><span class="co">#&gt; 41:        239        4     -48.9 13231.0           13279.9</span></span>
<span><span class="co">#&gt; 42:        142        5     -48.9 13231.0           13279.9</span></span>
<span><span class="co">#&gt; 43:        167        6     -48.9 13231.0           13279.9</span></span>
<span><span class="co">#&gt; 44:        225        7     -48.9 13231.0           13279.9</span></span>
<span><span class="co">#&gt; 45:       1599        8     -48.9 13231.0           13279.9</span></span>
<span><span class="co">#&gt;     blocksizes exon_num min_start max_end transcript_length</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt;       chr    start      end strand transcript_name rescaled_length</span></span>
<span><span class="co">#&gt;  1: chr10 87864548 87894025      +        PTEN-201          2947.7</span></span>
<span><span class="co">#&gt;  2: chr10 87894109 87925513      +        PTEN-201          3140.4</span></span>
<span><span class="co">#&gt;  3: chr10 87925557 87931046      +        PTEN-201           548.9</span></span>
<span><span class="co">#&gt;  4: chr10 87931089 87933013      +        PTEN-201           192.4</span></span>
<span><span class="co">#&gt;  5: chr10 87933251 87952118      +        PTEN-201          1886.7</span></span>
<span><span class="co">#&gt;  6: chr10 87952259 87957853      +        PTEN-201           559.4</span></span>
<span><span class="co">#&gt;  7: chr10 87958019 87960894      +        PTEN-201           287.5</span></span>
<span><span class="co">#&gt;  8: chr10 87961118 87965287      +        PTEN-201           416.9</span></span>
<span><span class="co">#&gt;  9: chr10 87864548 87894025      +        PTEN-204          2947.7</span></span>
<span><span class="co">#&gt; 10: chr10 87894109 87925513      +        PTEN-204          3140.4</span></span>
<span><span class="co">#&gt; 11: chr10 87925557 87931046      +        PTEN-204           548.9</span></span>
<span><span class="co">#&gt; 12: chr10 87931089 87933013      +        PTEN-204           192.4</span></span>
<span><span class="co">#&gt; 13: chr10 87933251 87952118      +        PTEN-204          1886.7</span></span>
<span><span class="co">#&gt; 14: chr10 87952259 87957853      +        PTEN-204           559.4</span></span>
<span><span class="co">#&gt; 15: chr10 87958019 87960894      +        PTEN-204           287.5</span></span>
<span><span class="co">#&gt; 16: chr10 87863113 87864454      +        PTEN-209           134.1</span></span>
<span><span class="co">#&gt; 17: chr10 87864548 87894025      +        PTEN-209          2947.7</span></span>
<span><span class="co">#&gt; 18: chr10 87894109 87925513      +        PTEN-209          3140.4</span></span>
<span><span class="co">#&gt; 19: chr10 87925557 87931046      +        PTEN-209           548.9</span></span>
<span><span class="co">#&gt; 20: chr10 87931089 87933013      +        PTEN-209           192.4</span></span>
<span><span class="co">#&gt; 21: chr10 87933251 87952118      +        PTEN-209          1886.7</span></span>
<span><span class="co">#&gt; 22: chr10 87952259 87957853      +        PTEN-209           559.4</span></span>
<span><span class="co">#&gt; 23: chr10 87958019 87960894      +        PTEN-209           287.5</span></span>
<span><span class="co">#&gt; 24: chr10 87961118 87965287      +        PTEN-209           416.9</span></span>
<span><span class="co">#&gt; 25: chr10 87864103 87864105      +        PTEN-211             0.2</span></span>
<span><span class="co">#&gt; 26: chr10 87864548 87894025      +        PTEN-211          2947.7</span></span>
<span><span class="co">#&gt; 27: chr10 87894109 87925513      +        PTEN-211          3140.4</span></span>
<span><span class="co">#&gt; 28: chr10 87925557 87931046      +        PTEN-211           548.9</span></span>
<span><span class="co">#&gt; 29: chr10 87931089 87933013      +        PTEN-211           192.4</span></span>
<span><span class="co">#&gt; 30: chr10 87933251 87952118      +        PTEN-211          1886.7</span></span>
<span><span class="co">#&gt; 31: chr10 87952259 87957853      +        PTEN-211           559.4</span></span>
<span><span class="co">#&gt; 32: chr10 87958019 87960894      +        PTEN-211           287.5</span></span>
<span><span class="co">#&gt; 33: chr10 87961118 87965287      +        PTEN-211           416.9</span></span>
<span><span class="co">#&gt; 34: chr10 87864548 87894025      +        PTEN-212          2947.7</span></span>
<span><span class="co">#&gt; 35: chr10 87894109 87931046      +        PTEN-212          3693.7</span></span>
<span><span class="co">#&gt; 36: chr10 87931089 87933013      +        PTEN-212           192.4</span></span>
<span><span class="co">#&gt; 37: chr10 87933251 87952118      +        PTEN-212          1886.7</span></span>
<span><span class="co">#&gt; 38: chr10 87952259 87957853      +        PTEN-212           559.4</span></span>
<span><span class="co">#&gt; 39: chr10 87958019 87960894      +        PTEN-212           287.5</span></span>
<span><span class="co">#&gt; 40: chr10 87961118 87965287      +        PTEN-212           416.9</span></span>
<span><span class="co">#&gt;       chr    start      end strand transcript_name rescaled_length</span></span></code></pre></div>
<p>We will first need to load our leafcutter junctions that we obtained
by running Regtools extract junctions on our BAM files. You can use one
of our preloaded cell types or first run this on your own BAM file and
then use as input for this function. We will look at junctions in hESC
data. We will first need to run ‘minicutter’ to cluster the junction
coordinates and obtain intron cluster events.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># load junctions for cell type of interest or input your own</span></span>
<span><span class="va">junctions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"hESC-MKNK2-G1_v41_basic.junc"</span>, package<span class="op">=</span><span class="st">"isoviz"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># run minicutter to get clusters </span></span>
<span><span class="va">intron_clusts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/isoviz_minicutter.html">isoviz_minicutter</a></span><span class="op">(</span>juncs_file <span class="op">=</span> <span class="va">junctions</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Printing summary of intron-cluster sizes = how many junctions across in clusters"</span></span></code></pre></div>
<p><img src="isoviz_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<pre><code><span><span class="co">#&gt; [1] "Printing summary of intron-cluster sizes = how many junctions across in clusters"</span></span></code></pre>
<p><img src="isoviz_files/figure-html/unnamed-chunk-5-2.png" width="700"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">intron_clusts</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 10</span></span></span>
<span><span class="co">#&gt;   chrom strand  start    end name        readcount idx_start idx_end cluster_idx</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> chr1  +       <span style="text-decoration: underline;">16</span>717 <span style="text-decoration: underline;">187</span>231 JUNC000000…         6         1       5           1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> chr1  +       <span style="text-decoration: underline;">23</span>782 <span style="text-decoration: underline;">194</span>308 JUNC000000…         3         2       6           2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> chr1  -       <span style="text-decoration: underline;">90</span>050  <span style="text-decoration: underline;">90</span>287 JUNC000000…         1     <span style="text-decoration: underline;">17</span>151   <span style="text-decoration: underline;">17</span>152        <span style="text-decoration: underline;">7</span>295</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> chr1  -      <span style="text-decoration: underline;">136</span>463 <span style="text-decoration: underline;">136</span>562 JUNC000000…         8     <span style="text-decoration: underline;">17</span>153   <span style="text-decoration: underline;">17</span>155        <span style="text-decoration: underline;">7</span>296</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> chr1  -      <span style="text-decoration: underline;">136</span>512 <span style="text-decoration: underline;">136</span>904 JUNC000000…         2     <span style="text-decoration: underline;">17</span>154   <span style="text-decoration: underline;">17</span>159        <span style="text-decoration: underline;">7</span>297</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> chr1  -      <span style="text-decoration: underline;">136</span>708 <span style="text-decoration: underline;">136</span>904 JUNC000000…         1     <span style="text-decoration: underline;">17</span>156   <span style="text-decoration: underline;">17</span>159        <span style="text-decoration: underline;">7</span>297</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: usage_ratio &lt;dbl&gt;</span></span></span></code></pre></div>
<p>Let’s take a look at how junctions were grouped into intron clusters.
Note, some of these clusters will only contain one junction (singleton).
You will have the option to change this when running
isoviz_minicutter.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For now we will use the expanded intron dataset with additional annotations by Megan </span></span>
<span><span class="co"># We will need an additional function to add these annotations </span></span>
<span><span class="va">intron_annotations</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"gencode_intron_all_data.rda"</span>, package<span class="op">=</span><span class="st">"isoviz"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="va">intron_annotations</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gencode_intron_all_data</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 10</span></span></span>
<span><span class="co">#&gt;   chr   junc_start junc_end gene_id           strand junc_id gene_name gene_type</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> chr1       <span style="text-decoration: underline;">30</span>039    <span style="text-decoration: underline;">30</span>564 ENSG00000243485.5 +      junc1   MIR1302-… lncRNA   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> chr1       <span style="text-decoration: underline;">30</span>667    <span style="text-decoration: underline;">30</span>976 ENSG00000243485.5 +      junc2   MIR1302-… lncRNA   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> chr1       <span style="text-decoration: underline;">35</span>174    <span style="text-decoration: underline;">35</span>277 ENSG00000237613.2 -      junc3   FAM138A   lncRNA   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> chr1       <span style="text-decoration: underline;">35</span>481    <span style="text-decoration: underline;">35</span>721 ENSG00000237613.2 -      junc4   FAM138A   lncRNA   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> chr1       <span style="text-decoration: underline;">65</span>433    <span style="text-decoration: underline;">65</span>520 ENSG00000186092.7 +      junc5   OR4F5     protein_…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> chr1       <span style="text-decoration: underline;">65</span>573    <span style="text-decoration: underline;">69</span>037 ENSG00000186092.7 +      junc6   OR4F5     protein_…</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 2 more variables: junction_category &lt;chr&gt;, transcript_isoforms &lt;chr&gt;</span></span></span></code></pre></div>
<p>Let’s continue working on RBFOX2. Now we have to map the observed
junctions with their corresponding exons and transcript isoforms. To do
this, we will use the <code>isoviz_map_junctions</code> function. Make
sure to check strand of gene!</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mapped_junctions</span> <span class="op">=</span> <span class="fu"><a href="../reference/isoviz_map_junctions.html">isoviz_map_junctions</a></span><span class="op">(</span>cell_type <span class="op">=</span> <span class="st">"hESC"</span>, <span class="va">rbfox2_introns</span>, <span class="va">intron_clusts</span>, <span class="va">gencode_intron_all_data</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "removing junction IDs obtained from clustering introns, using default names from gencode file"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mapped_junctions</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 24</span></span></span>
<span><span class="co">#&gt;   chr   intron_starts intron_ends strand junc.counts idx_start idx_end</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> chr10      87<span style="text-decoration: underline;">893</span>773    87<span style="text-decoration: underline;">894</span>025 +                1     <span style="text-decoration: underline;">37</span>401   <span style="text-decoration: underline;">37</span>402</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> chr10      87<span style="text-decoration: underline;">864</span>548    87<span style="text-decoration: underline;">894</span>025 +               63     <span style="text-decoration: underline;">37</span>397   <span style="text-decoration: underline;">37</span>402</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> chr10      87<span style="text-decoration: underline;">864</span>548    87<span style="text-decoration: underline;">894</span>025 +               63     <span style="text-decoration: underline;">37</span>397   <span style="text-decoration: underline;">37</span>402</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> chr10      87<span style="text-decoration: underline;">864</span>548    87<span style="text-decoration: underline;">894</span>025 +               63     <span style="text-decoration: underline;">37</span>397   <span style="text-decoration: underline;">37</span>402</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> chr10      87<span style="text-decoration: underline;">864</span>548    87<span style="text-decoration: underline;">894</span>025 +               63     <span style="text-decoration: underline;">37</span>397   <span style="text-decoration: underline;">37</span>402</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> chr10      87<span style="text-decoration: underline;">864</span>548    87<span style="text-decoration: underline;">894</span>025 +               63     <span style="text-decoration: underline;">37</span>397   <span style="text-decoration: underline;">37</span>402</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 17 more variables: cluster_idx &lt;dbl&gt;, usage_ratio &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   cluster.counts &lt;dbl&gt;, gene_id &lt;chr&gt;, trans_id &lt;chr&gt;, gene_name &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   transcript_name &lt;chr&gt;, gene_type &lt;chr&gt;, transcript_type &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   junction_category &lt;chr&gt;, junc_id &lt;chr&gt;, transcript_isoforms &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   junc.per.cluster &lt;dbl&gt;, junc.usage &lt;dbl&gt;, cell_line &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   transcript_targetable &lt;chr&gt;, isoform_expressed &lt;chr&gt;</span></span></span></code></pre></div>
<p>Now we are ready to make a plot!</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/isoviz_plot_juncs_to_iso.html">isoviz_plot_juncs_to_iso</a></span><span class="op">(</span><span class="va">mapped_junctions</span>, <span class="va">rbfox2_exons</span>,</span>
<span>                                    cell_type <span class="op">=</span> <span class="st">"hESC"</span>,</span>
<span>                                    junc_usage <span class="op">=</span> <span class="fl">5</span>, <span class="co">#min junc usage to be included </span></span>
<span>                                    include_all_juncs <span class="op">=</span> <span class="cn">TRUE</span>, intron_scale <span class="op">=</span> <span class="st">"no"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Adding missing grouping variables: `cluster_idx`</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(chr, intron_starts, intron_ends, strand,</span></span>
<span><span class="co">#&gt; junc.counts, cluster_idx, junction_category, junc_id, junc.per.cluster,</span></span>
<span><span class="co">#&gt; junc.usage)`</span></span></code></pre></div>
<p><img src="isoviz_files/figure-html/unnamed-chunk-8-1.png" width="700">
Now let’s make the same plot but with scaled introns so that we can see
the exons better! Note, this currently only works for junctions that we
are able to annotate to transcript isoforms in the top plot!</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/isoviz_plot_juncs_to_iso.html">isoviz_plot_juncs_to_iso</a></span><span class="op">(</span><span class="va">mapped_junctions</span>, <span class="va">rbfox2_exons</span>, <span class="va">rbfox2_introns</span>,</span>
<span>                                    cell_type <span class="op">=</span> <span class="st">"hESC"</span>,</span>
<span>                                    junc_usage <span class="op">=</span> <span class="fl">50</span>, <span class="co">#min junc usage to be included </span></span>
<span>                                    intron_scale <span class="op">=</span> <span class="st">"yes"</span>, intron_scale_width <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                                    include_all_juncs <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "+"</span></span>
<span><span class="co">#&gt; [1] "PTEN-209"</span></span>
<span><span class="co">#&gt; [1] "getting ref coordinates"</span></span>
<span><span class="co">#&gt; [1] "PTEN-201"</span></span>
<span><span class="co">#&gt; [1] "PTEN-204"</span></span>
<span><span class="co">#&gt; [1] "PTEN-211"</span></span>
<span><span class="co">#&gt; [1] "PTEN-212"</span></span>
<span><span class="co">#&gt; [1] 1</span></span>
<span><span class="co">#&gt; Adding missing grouping variables: `cluster_idx`</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(chr, intron_starts, intron_ends, strand,</span></span>
<span><span class="co">#&gt; junc.counts, cluster_idx, junction_category, junc_id, junc.per.cluster,</span></span>
<span><span class="co">#&gt; junc.usage)`</span></span></code></pre></div>
<p><img src="isoviz_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/isoviz_plot_juncs_to_iso.html">isoviz_plot_juncs_to_iso</a></span><span class="op">(</span><span class="va">mapped_junctions</span>, <span class="va">rbfox2_exons</span>, <span class="va">rbfox2_introns</span>,</span>
<span>                                    cell_type <span class="op">=</span> <span class="st">"hESC"</span>,</span>
<span>                                    junc_usage <span class="op">=</span> <span class="fl">5</span>, <span class="co">#min junc usage to be included </span></span>
<span>                                    intron_scale <span class="op">=</span> <span class="st">"yes"</span>, intron_scale_width <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                                    include_all_juncs <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                     include_specific_junctions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"junc178144"</span>, <span class="st">"junc178139"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "+"</span></span>
<span><span class="co">#&gt; [1] "PTEN-209"</span></span>
<span><span class="co">#&gt; [1] "getting ref coordinates"</span></span>
<span><span class="co">#&gt; [1] "PTEN-201"</span></span>
<span><span class="co">#&gt; [1] "PTEN-204"</span></span>
<span><span class="co">#&gt; [1] "PTEN-211"</span></span>
<span><span class="co">#&gt; [1] "PTEN-212"</span></span>
<span><span class="co">#&gt; [1] 1</span></span>
<span><span class="co">#&gt; Adding missing grouping variables: `cluster_idx`</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(chr, intron_starts, intron_ends, strand,</span></span>
<span><span class="co">#&gt; junc.counts, cluster_idx, junction_category, junc_id, junc.per.cluster,</span></span>
<span><span class="co">#&gt; junc.usage)`</span></span></code></pre></div>
<p><img src="isoviz_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by .</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
